modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManagerImpl.java

```diff
@@ -374,7 +375,44 @@ private void truncateUpTo(Catalog catalog) {
 
     private CompletableFuture<Integer> saveUpdateAndWaitForActivation(UpdateProducer updateProducer) {
         return saveUpdate(updateProducer, 0)
-                .thenCompose(newVersion -> {
+                .handle((newVersion, err) -> {
+                    if (err != null) {
+                        Throwable error;
+                        Catalog catalog;
+
+                        Throwable errUnwrapped = err instanceof CompletionException ? err.getCause() : err;
+
+                        if (errUnwrapped instanceof CatalogVersionAwareValidationException) {
+                            CatalogVersionAwareValidationException err0 = (CatalogVersionAwareValidationException) errUnwrapped;
+                            catalog = catalogByVer.get(err0.version());
+                            error = err0.initial();
+                        } else {
+                            catalog = catalogByVer.lastEntry().getValue();
+                            error = err;
+                        }
+
+                        if (catalog.version() == 0) {
+                            return failedFuture(error).thenApply(unused -> newVersion);
+                        }
+
+                        HybridTimestamp tsSafeForRoReadingInPastOptimization =
+                                clusterWideEnsuredActivationTsSafeForRoReads(
```

Comment:
How about extracting this expression to a method having one parameter of type `Catalog`? It's pretty complex and it's used twice.

Reply:
done